designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,324 @@
+* Ability to specify client certificates from external files in standard [formats](https://en.wikipedia.org/wiki/X.509#Certificate_filename_extensions)
+* Ability to configure settings with command line
+* Each source can have at most one client certificate configuration
+* If client certificate configuration returns several valid x509 certificates (storeCert for example) first instance will be used
+* Client certificate configurations can be overwritten in common way
+
+## Solution
+
+1) Find specified in NuGet configuration file certificates (common implementation for [all products](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Configuration) )
+2) Apply them prior any web request to appropriate http handler (product specific)
+
+## NuGet configuration changes
+
+Provide new configuration section `clientCerts` which may have children of 2 types:
+
+1. `storeCert` - certificate import from [certificate store](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores). Internally uses [X509Certificate2Collection.Find](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.find?view=netframework-4.8#System_Security_Cryptography_X509Certificates_X509Certificate2Collection_Find_System_Security_Cryptography_X509Certificates_X509FindType_System_Object_System_Boolean_) method.
+Can be configured with:
+    - Attribute `packageSource`. Required. `packageSources` source name reference.
+    - Attribute `storeLocation`. [Possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storelocation?view=netframework-4.8#fields). Optional. Equals `CurrentUser` by default.
+    - Attribute `storeName`. [Possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename?view=netframework-4.8#fields). Optional. Equals `My` by default.
+    - Attribute `findBy`. [Possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.8#fields) with out FindBy prefix. Optional. Equals `thumbprint` by default.
+    - Attribute `findValue`. `value` - the search criteria as an object. Required.
+
+2. `fileCert` - certificate import from file (DER encoded x.509, Base-64 encoded x.509, PKCS)
+Can be configured with:
+    - Attribute `packageSource`. Required. `packageSources` source name reference.
+    - Attribute `path`. Absolute or relative path to certificate file. Required.
+    - Attribute `password`. Plain or encrypted password string. Optional. Encrypted in same manner as [PackageSourceCredential](https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#packagesourcecredentials) password.
```

Comment:
For credentials, we have different keys for encrypted and plain text passwords. If we use the same field here, how do we know if it's encrypted or plain text?  I think it will be better to have separate properties.

Reply:
done